### ISSUE TITLE
Websocket not working?

### ISSUE THREAD

**Issue by MyWay** [Brettdah/vaultwarden-traefiked#1]
Hi,
from the developer console I can see errors regarding websockets' connection:

```
Error: Failed to start the connection: Error: WebSocket failed to connect. The connection could not be found on the server, either the endpoint may not be a SignalR endpoint, the connection ID is not present on the server, or there is a proxy blocking WebSockets. If you have multiple servers check that sticky sessions are enabled.
log @ Utils.ts:193
_startInternal @ HttpConnection.ts:298
await in _startInternal (async)
start @ HttpConnection.ts:137
_startInternal @ HubConnection.ts:206
_startWithStateTransitions @ HubConnection.ts:180
start @ HubConnection.ts:167
(anonymous) @ notifications.service.ts:205
r @ main.5f8690f5c03a207c390a.js:54
t.invoke @ zone.js:409
onInvoke @ ng_zone.ts:418
t.invoke @ zone.js:408
e.run @ zone.js:169
(anonymous) @ zone.js:1326
t.invokeTask @ zone.js:443
onInvokeTask @ ng_zone.ts:403
t.invokeTask @ zone.js:442
e.runTask @ zone.js:214
b @ zone.js:632
Promise.then (async)
y @ zone.js:608
m @ zone.js:619
t.scheduleTask @ zone.js:433
onScheduleTask @ zone.js:320
t.scheduleTask @ zone.js:423
e.scheduleTask @ zone.js:257
e.scheduleMicroTask @ zone.js:277
C @ zone.js:1316
P @ zone.js:1253
(anonymous) @ zone.js:1169
(anonymous) @ zone.js:1185
Promise.then (async)
(anonymous) @ zone.js:1577
t @ zone.js:1351
B.t.then @ zone.js:1576
P @ zone.js:1220
(anonymous) @ zone.js:1169
(anonymous) @ zone.js:1185
(anonymous) @ main.5f8690f5c03a207c390a.js:54
t @ zone.js:1351
l @ main.5f8690f5c03a207c390a.js:54
r @ main.5f8690f5c03a207c390a.js:54
t.invoke @ zone.js:409
onInvoke @ ng_zone.ts:418
t.invoke @ zone.js:408
e.run @ zone.js:169
(anonymous) @ zone.js:1326
t.invokeTask @ zone.js:443
onInvokeTask @ ng_zone.ts:403
t.invokeTask @ zone.js:442
e.runTask @ zone.js:214
b @ zone.js:632
Promise.then (async)
y @ zone.js:608
m @ zone.js:619
t.scheduleTask @ zone.js:433
onScheduleTask @ zone.js:320
t.scheduleTask @ zone.js:423
e.scheduleTask @ zone.js:257
e.scheduleMicroTask @ zone.js:277
C @ zone.js:1316
P @ zone.js:1253
(anonymous) @ zone.js:1169
(anonymous) @ zone.js:1185
Promise.then (async)
(anonymous) @ zone.js:1577
t @ zone.js:1351
B.t.then @ zone.js:1576
P @ zone.js:1220
(anonymous) @ zone.js:1169
(anonymous) @ zone.js:1185
l @ main.5f8690f5c03a207c390a.js:54
r @ main.5f8690f5c03a207c390a.js:54
t.invoke @ zone.js:409
onInvoke @ ng_zone.ts:418
t.invoke @ zone.js:408
e.run @ zone.js:169
(anonymous) @ zone.js:1326
t.invokeTask @ zone.js:443
onInvokeTask @ ng_zone.ts:403
t.invokeTask @ zone.js:442
e.runTask @ zone.js:214
b @ zone.js:632
e.invokeTask @ zone.js:529
b @ zone.js:1727
w @ zone.js:1758
x @ zone.js:1794
postMessage (async)
U @ VM83 webauthn.b50e3a9527b3e4de41eb.js:1
Promise.then (async)
h @ VM83 webauthn.b50e3a9527b3e4de41eb.js:1
m @ VM83 webauthn.b50e3a9527b3e4de41eb.js:1
(anonymous) @ VM83 webauthn.b50e3a9527b3e4de41eb.js:1
zone.js:1109 Unhandled Promise rejection: WebSocket failed to connect. The connection could not be found on the server, either the endpoint may not be a SignalR endpoint, the connection ID is not present on the server, or there is a proxy blocking WebSockets. If you have multiple servers check that sticky sessions are enabled. ; Zone: <root> ; Task: null ; Value: Error: WebSocket failed to connect. The connection could not be found on the server, either the endpoint may not be a SignalR endpoint, the connection ID is not present on the server, or there is a proxy blocking WebSockets. If you have multiple servers check that sticky sessions are enabled
```

From the traefik logs I can see I hit a 502 bad gateway, I was curious if that happens to you, maybe due to changes in some newer version?

**Comment by Brettdah** [comment]
Hello, I will look, but it seems either we don't need the websockets or at least I don't use them cause to me every action I do seems to work properly... I'm not using, the extention, nor the web site of vaultwarden, just the app on Windows, Linux, and Android

[EDIT]
![image](https://user-images.githubusercontent.com/18658617/215317688-83db20ac-ca8e-4279-bd45-21950361b2ea.png)
It seems I have the same issue but so far I did not noticed any problems using the apps... I will still investiguate to fix that.

I'll gladly accept a PR if you are to find the fix first ;) just be sure not to disclose your domain I will double check that anyway ;)

[EDIT2]
I went to the [vaultwarden doc](https://github.com/dani-garcia/vaultwarden/wiki/Proxy-examples#migrated-labels-plus-http-to-https-redirect) it seems I followed it to the letter...

**Comment by MyWay** [comment]
Yes, it seems the problem is when you have to sync among browsers and/or apps, e.g. if you open your account from a browser tab and another one from an anonymous tab and you add some data, data don't get synced and you risk to loss data. 

Yeah, I also looked at the official docs and couldn't find anything, that's why I was asking if it happened to you too!

I'll try to post it on the official repo. Thanks.

**Comment by Brettdah** [comment]
You're welcome, I'll try some things on my side too.

**Comment by Brettdah** [comment]
just saw something [here](https://github.com/dani-garcia/vaultwarden/wiki/Enabling-WebSocket-notifications) worth mentionning :
if ever it's not in the config already, we may need to make sure to pass the **Connection** and **Upgrade** headers

**Comment by MyWay** [comment]
In my case I had to add an internal network (I was exposing only the "public" one); so, probably, when it tried to connect to the internal address it could not reach it.

```
networks:
      - traefik-public
      - traefik-internal
```

and then on the labels I added:
`      - "traefik.docker.network=traefik-public"`

Then the console didn't report the error anymore, on my network tab I had a 101 message and testing with two browsers worked as expected.

**Comment by Brettdah** [comment]
hi,
Let's see If I understood that right :
your compose file would look like that :

``` yaml
version: '3.1'
# NETWORKS
networks:
  proxy:
    external: true
  proxy-internal:

#SERVICES
services:
  bitwarden:
    image: vaultwarden/server:1.27.0-alpine
    deploy:
      labels:
        - traefik.enable=true
        - traefik.docker.network=proxy
        - traefik.http.middlewares.redirect-https.redirectScheme.scheme=https
        - traefik.http.middlewares.redirect-https.redirectScheme.permanent=true
        - traefik.http.routers.bitwarden-ui-https.rule=Host(`vault.home.tux4all.fr`)
        - traefik.http.routers.bitwarden-ui-https.entrypoints=https
        - traefik.http.routers.bitwarden-ui-https.tls=true
        - traefik.http.routers.bitwarden-ui-https.service=bitwarden-ui
        - traefik.http.routers.bitwarden-ui-http.rule=Host(`vault.home.tux4all.fr`)
        - traefik.http.routers.bitwarden-ui-http.entrypoints=http
        - traefik.http.routers.bitwarden-ui-http.middlewares=redirect-https
        - traefik.http.routers.bitwarden-ui-http.service=bitwarden-ui
        - traefik.http.services.bitwarden-ui.loadbalancer.server.port=80
        - traefik.http.routers.bitwarden-websocket-https.rule=Host(`vault.home.tux4all.fr`) && Path(`/notifications/hub`)
        - traefik.http.routers.bitwarden-websocket-https.entrypoints=https
        - traefik.http.routers.bitwarden-websocket-https.tls=true
        - traefik.http.routers.bitwarden-websocket-https.service=bitwarden-websocket
        - traefik.http.routers.bitwarden-websocket-http.rule=Host(`vault.home.tux4all.fr`) && Path(`/notifications/hub`)
        - traefik.http.routers.bitwarden-websocket-http.entrypoints=http
        - traefik.http.routers.bitwarden-websocket-http.middlewares=redirect-https
        - traefik.http.routers.bitwarden-websocket-http.service=bitwarden-websocket
        - traefik.http.services.bitwarden-websocket.loadbalancer.server.port=3012
    networks:
      - proxy
      - proxy-internal
    environment:
      - SIGNUPS_ALLOWED="false"
      - WEBSOCKET_ENABLED="true"
      - LOG_FILE="/var/log/vaultwarden/vaultwarden.log"
      - LOG_LEVEL=debug
      - TZ="Europe/Paris"
    volumes:
      - /etc/timezone:/etc/timezone:ro
      - vaultwarden-logs:/var/log/vaultwarden
      # Data storage AKA DO NOT COMMENT THAT YOU ASS HOLE !!! yep I experienced it ^^
      - $PWD/data/:/data
```
I'm keeping the name of my networks ;)
but will you  just do that or add it to traefik too ?

**Comment by MyWay** [comment]
Yes, this way seems vaultwarden is able to reach 10.x.x.x ip using websockets.

**Comment by Brettdah** [comment]
ok and you have had the **proxy-internal** to traefik to right ?
If you have a source/doc describing the configuraiton I would like to read it :)

**Comment by MyWay** [comment]
Using your config it should be:
```
version: '3'
# NETWORKS
networks:
  proxy:
    external: true
  internal:
    external: true

#SERVICES
services:
  bitwarden:
    image: vaultwarden/server:latest
    security_opt:
      - no-new-privileges:true
    networks:
      - proxy
      - internal
    environment:
      # After you setted up your user(s)
      - "SIGNUPS_ALLOWED=false"
    labels:
      - traefik.enable=true
      - traefik.docker.network=proxy
      # Redirect to get the certs
      - traefik.http.middlewares.redirect-https.redirectScheme.scheme=https
      - traefik.http.middlewares.redirect-https.redirectScheme.permanent=true
      # routers & services
      - traefik.http.routers.bitwarden-ui-https.rule=Host(`sub.my-domain.tld`)
      - traefik.http.routers.bitwarden-ui-https.entrypoints=https
      - traefik.http.routers.bitwarden-ui-https.tls=true
      - traefik.http.routers.bitwarden-ui-https.service=bitwarden-ui
      - traefik.http.routers.bitwarden-ui-http.rule=Host(`sub.my-domain.tld`)
      - traefik.http.routers.bitwarden-ui-http.entrypoints=http
      - traefik.http.routers.bitwarden-ui-http.middlewares=redirect-https
      - traefik.http.routers.bitwarden-ui-http.service=bitwarden-ui
      - traefik.http.services.bitwarden-ui.loadbalancer.server.port=80
      - traefik.http.routers.bitwarden-websocket-https.rule=Host(`sub.my-domain.tld`) && Path(`/notifications/hub`)
      - traefik.http.routers.bitwarden-websocket-https.entrypoints=https
      - traefik.http.routers.bitwarden-websocket-https.tls=true
      - traefik.http.routers.bitwarden-websocket-https.service=bitwarden-websocket
      - traefik.http.routers.bitwarden-websocket-http.rule=Host(`sub.my-domain.tld`) && Path(`/notifications/hub`)
      - traefik.http.routers.bitwarden-websocket-http.entrypoints=http
      - traefik.http.routers.bitwarden-websocket-http.middlewares=redirect-https
      - traefik.http.routers.bitwarden-websocket-http.service=bitwarden-websocket
      - traefik.http.services.bitwarden-websocket.loadbalancer.server.port=3012

    volumes:
      # Stockage des fichiers Let's encrypt
      - ./data/:/data
```
with the internal network created as such, because I'm on docker swarm: https://docs.docker.com/engine/reference/commandline/network_create/#internal

My traefik service does not use the internal network, only the public (your proxy) one. I use the internal one for services that don't have to communicate externally, or I use both networks when both the service needs internal and external communication. It's still not clear to me why vaultwarden needs the internal network though.

Didn't have time to test, let me know if it works for you and sorry for the delay.

**Comment by Brettdah** [comment]
Fix tested with upgrade 1.29.1 > not working opening #3
